### PR TITLE
Pass all graphdriver create() parameters in a struct

### PR DESCRIPTION
**Pull request by rhvgoyal**
This allows for easy extension of adding more parameters to existing
parameters list. Otherwise adding a single parameter changes code
at so many places.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

### CODE REVIEW

**Comment by rhvgoyal** [comment]
cc @crosbymichael @thaJeztah

**Comment by thaJeztah** [comment]
is this going to affect the experimental graphdriver-plugins? https://github.com/docker/docker/blob/master/experimental/plugins_graphdriver.md

**Comment by rhvgoyal** [comment]
@thaJeztah I don't think that graphdriver plugins will see any affect. All the changes are internal.

**Comment by dmcgowan** [comment]
Having the non-optional parameters outside of the options struct is a much cleaner pattern. I think this would include both the id and the parent and would end up looking more like...

```
func (a *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) error
```

What is nice about that approach is in most trivial cases nil could be passed in for the opts. The way it is coded up now a pointer is passed in but passing in nil is never allowed. Since that would be a large update I will let other maintainers weigh in before asking to update :)

**Comment by rhvgoyal** [comment]
`parent` is optional parameter as well.

**Comment by crosbymichael** [comment]
Moving the id out like @dmcgowan said makes sense to me.  If parent is optional then it would be best to leave in the opts struct

**Comment by dmcgowan** [comment]
I am fine putting parent in options struct. Parent can be seen as optional or `""` can be viewed as scratch, doesn't matter. Was just suggesting having it outside since it is commonly given throughout the tests while the other options will usually be empty.

**Comment by rhvgoyal** [comment]
Ok, I am fine either way. Please decide if you want `parent` inside struct or outside. For now, I will start with putting it outside struct. Do let me know if you don't like it.

**Comment by crosbymichael** [comment]
@rhvgoyal that is fine with me

**Comment by dmcgowan** [comment]
> For now, I will start with putting it outside struct.

:+1:

**Comment by rhvgoyal** [comment]
Moved `id` and `parent` fields outside `CreateOpts` struct.

File: daemon/graphdriver/driver.go [hunk 1]
```diff
@@ -50,10 +57,10 @@ type ProtoDriver interface {
 	String() string
 	// CreateReadWrite creates a new, empty filesystem layer that is ready
 	// to be used as the storage for a container.
-	CreateReadWrite(id, parent, mountLabel string, storageOpt map[string]string) error
+	CreateReadWrite(id, parent string, opts *CreateOpts) error
 	// Create creates a new, empty, filesystem layer with the
 	// specified id and parent and mountLabel. Parent and mountLabel may be "".
```
Comment (dmcgowan) [inline daemon/graphdriver/driver.go]: Update comment to mention create options instead of mount label and mention that the create options may be nil.

**Comment by dmcgowan** [comment]
Code looks good, just one comment needs updating

**Comment by crosbymichael** [comment]
LGTM

**Comment by rhvgoyal** [comment]
@dmcgowan Updated patch to address the comment. PTAL.

**Comment by dmcgowan** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
if it looks good to @dmcgowan, it LGTM

**Pull request merged**